pubsub/producer.go

```diff
@@ -0,0 +1,209 @@
+				acked[msg.ID] = msg.Value
+			}
+			for k, v := range acked {
+				// Only re-insert messages that were removed the the pending list first.
+				_, err := p.reproduce(ctx, v, k)
+				if err != nil {
+					log.Error("Re-inserting pending messages with inactive consumers", "error", err)
+				}
+			}
+			return p.cfg.CheckPendingInterval
+		},
+	)
+	// Iteratively check whether result were returned for some queries.
+	p.StopWaiter.CallIteratively(func(ctx context.Context) time.Duration {
+		p.promisesLock.Lock()
+		defer p.promisesLock.Unlock()
+		for id, promise := range p.promises {
+			res, err := p.client.Get(ctx, id).Result()
+			if err != nil {
+				if errors.Is(err, redis.Nil) {
+					continue
+				}
+				log.Error("Error reading value in redis", "key", id, "error", err)
+			}
+			var tmp T
+			val, err := tmp.Unmarshal(res)
+			if err != nil {
+				log.Error("Error unmarshaling", "value", res, "error", err)
+				continue
+			}
+			promise.Produce(val)
+			delete(p.promises, id)
+		}
+		return p.cfg.CheckResultInterval
+	})
+}
+
+// reproduce is used when Producer claims ownership on the pending
+// message that was sent to inactive consumer and reinserts it into the stream,
+// so that seamlessly return the answer in the same promise.
+func (p *Producer[T]) reproduce(ctx context.Context, value T, oldKey string) (*containers.Promise[T], error) {
+	id, err := p.client.XAdd(ctx, &redis.XAddArgs{
+		Stream: p.cfg.RedisStream,
+		Values: map[string]any{messageKey: value.Marshal()},
+	}).Result()
+	if err != nil {
+		return nil, fmt.Errorf("adding values to redis: %w", err)
+	}
+	p.promisesLock.Lock()
+	defer p.promisesLock.Unlock()
+	promise := p.promises[oldKey]
+	if oldKey == "" || promise == nil {
+		p := containers.NewPromise[T](nil)
+		promise = &p
+	}
+	p.promises[id] = promise
+	return promise, nil
+}
+
+func (p *Producer[T]) Produce(ctx context.Context, value T) (*containers.Promise[T], error) {
+	return p.reproduce(ctx, value, "")
+}
+
+// Check if a consumer is with specified ID is alive.
+func (p *Producer[T]) isConsumerAlive(ctx context.Context, consumerID string) bool {
+	val, err := p.client.Get(ctx, heartBeatKey(consumerID)).Int64()
+	if err != nil {
+		return false
+	}
+	return time.Now().UnixMilli()-val < int64(p.cfg.KeepAliveTimeout.Milliseconds())
+}
+
+func (p *Producer[T]) checkPending(ctx context.Context) ([]*Message[T], error) {
+	pendingMessages, err := p.client.XPendingExt(ctx, &redis.XPendingExtArgs{
+		Stream: p.cfg.RedisStream,
+		Group:  p.cfg.RedisGroup,
+		Start:  "-",
+		End:    "+",
+		Count:  100,
+	}).Result()
+
+	if err != nil && !errors.Is(err, redis.Nil) {
+		return nil, fmt.Errorf("querying pending messages: %w", err)
+	}
+	if len(pendingMessages) == 0 {
+		return nil, nil
+	}
+	// IDs of the pending messages with inactive consumers.
+	var ids []string
+	inactive := make(map[string]bool)
+	for _, msg := range pendingMessages {
+		if !inactive[msg.Consumer] || p.isConsumerAlive(ctx, msg.Consumer) {
```

Comment:
a live consumer that is active and has 100 pending messages will be tested 100 times